src/server/protocol_client.cc

```diff
@@ -195,8 +195,13 @@ ProtocolClient::ProtocolClient(ServerContext context) : server_context_(std::mov
 }
 
 ProtocolClient::~ProtocolClient() {
+  // FIXME: We should close the socket explictly outside of the constructor. This currently
+  // breaks test_cancel_replication_immediately.
   if (sock_) {
-    auto ec = sock_->Close();
+    std::error_code ec;
+    sock_->proactor()->Await([this, &ec]() {
+      ec = sock_->Close();
+    });
```

Comment:
What's wrong with closing in destructor?

Reply:
@romange thinks it's a bad pattern (and calling `Await()` in a destructor _is_ a bit dangerous TBH)